Add functionality to Amber Scoop layer

ilkkasaa opened the pull request:
Intercept following commands:
-vkCmdBeginRenderPass
-vkCmdBindPipeline
-vkCmdDraw
-vkCmdDrawIndexed

Implements simple draw call handling and Amber file generator
that generates non-functional example Amber files.

ilkkasaa committed:
Add some functionality to Amber Scoop layer

Intercept following commands:
-vkCmdBeginRenderPass
-vkCmdBindPipeline
-vkCmdDraw
-vkCmdDrawIndexed

Implements simple draw call handling and Amber file generator
that generates non-functional example Amber files.

ilkkasaa commented:
I disabled `warning C5045: Compiler will insert Spectre mitigation for memory load if /Qspectre switch specified` because Windows debug build was failing (see https://github.com/google/gf-layers/runs/1081462800 for more info). According to MS docs, I think the warning can be safely disabled: https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/c5045?view=vs-2019

afd commented on src/VkLayer_GF_amber_scoop/include/vk_layer_amber_scoop/vk_deep_copy.h:
```diff
@@ -0,0 +1,50 @@
+namespace gf_layers::amber_scoop_layer {
+
+template <typename T>
+T* CopyArray(T const* p_data, uint32_t num_elements, uint32_t offset) {
```
I think we should have a comment for all functions in header files, even if it's fairly obvious what the function does.

afd replied:
... in particular it would be good to document that this performs memory allocation.

afd commented on src/VkLayer_GF_amber_scoop/include/vk_layer_amber_scoop/vk_deep_copy.h:
```diff
@@ -0,0 +1,50 @@
+}
+
+template <typename T>
+T* CopyArray(T const* p_data, uint32_t num_elements) {
```
Add comment.

afd commented on src/VkLayer_GF_amber_scoop/include/vk_layer_amber_scoop/vk_deep_copy.h:
```diff
@@ -0,0 +1,50 @@
+VkRenderPassBeginInfo DeepCopy(
+    VkRenderPassBeginInfo const* p_render_pass_begin_info);
+
+void DeepDelete(VkRenderPassBeginInfo const* p_render_pass_begin_info);
```
Add comments.

afd commented on src/VkLayer_GF_amber_scoop/include/vk_layer_amber_scoop/vulkan_commands.h:
```diff
@@ -0,0 +1,146 @@
+class CmdDraw;
+class CmdDrawIndexed;
+
+class Cmd {
```
A top-level comment for the class would be good, with an indication of why we capture particular kinds of commands, so that when someone is working on the code base in the future they'll know when it's appropriate to add more commands here.

afd commented on src/VkLayer_GF_amber_scoop/include/vk_layer_amber_scoop/vulkan_commands.h:
```diff
@@ -0,0 +1,146 @@
+
+  ~CmdBeginRenderPass() override { DeepDelete(&render_pass_begin_); }
+
+  VkRenderPassBeginInfo render_pass_begin_;
+  VkSubpassContents contents_;
+
+  // Disabled copy/move constructors and copy/move assign operators
+  CmdBeginRenderPass(const CmdBeginRenderPass&) = delete;
+  CmdBeginRenderPass(CmdBeginRenderPass&&) = delete;
+  CmdBeginRenderPass& operator=(const CmdBeginRenderPass&) = delete;
+  CmdBeginRenderPass& operator=(CmdBeginRenderPass&&) = delete;
```
We should follow Google C++ style.  Thus:

- [All class data members should be private](https://google.github.io/styleguide/cppguide.html#Access_Control)

- If we really don't like the idea of making these fields private then we should [make these things structs](https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes) -- but the style says "if in doubt, use a class".  (I prefer classes and private members.)

- We should follow the [recommended declaration order](https://google.github.io/styleguide/cppguide.html#Declaration_Order) and in particular have the data members after the methods.

afd replied:
Same comments apply to the other classes here.

paulthomson replied:
Looking at how these are used, I would be OK with making them structs. They are basically just deep copies of other structs anyway, and once you get access to one of the structs, you could start mutating it, so there is not much benefit from "hiding" our fields.

ilkkasaa replied:
I agree that these should be structs. One thing about field/parameter naming: if we use structs, then the field names shouldn't have the trailing underscore (https://google.github.io/styleguide/cppguide.html#Variable_Names) and that causes variable shadow warning in constructor initializer list. F.ex. `explicit Cmd(Kind kind) : kind(kind) {}` is accepted by the language (and works as expected), but causes warning `constructor parameter 'kind' shadows the field 'kind' of 'Cmd' [-Wshadow-field-in-constructor]`

So, should we keep the trailing underscore in the member variables to avoid this issue?
Also, looks like I forgot to change the parameter names to lowercase with underscore in this file.

paulthomson replied:
I would say don't keep the trailing underscore and then either:

 - Put a "_param" suffix on the constructor parameters, or
 - Use a pragma across the entire file.

```
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wshadow-field-in-constructor"
...
#pragma clang diagnostic pop
```

ilkkasaa replied:
Oh, well. I changed them to structs, but looks like clang-tidy treats structs as classes if there's any functions defined and therefore it will give a warning about public members.

afd commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -64,13 +83,15 @@ namespace {
 struct AmberScoopLayerSettings {
   bool init = false;
   uint64_t start_draw_call = 0;
-  uint64_t draw_call_count = 0;
-  std::string output_file_prefix;
+  uint64_t draw_call_count = 1;
+  uint64_t last_draw_call = 0;
+  std::string output_file_prefix = "amber_scoop_output";
```
Can we have comments on these fields to briefly explain what they are for and any relationships between them?

E.g. presumably start_draw_call must be <= last_draw_call is an invariant that needs to hold.  Is draw_call_count always last_draw_call - start_draw_call + 1?  If so, is it redundant?

According to Google Style, if invariants need to be maintained between fields then a class with private members should be used (to avoid being able to break an invariant by setting a public field).

paulthomson replied:
I agree that these should have comments. I am happy to leave this as a struct given that it is read-only once initialized and is basically just a blob of data. I would perhaps remove `draw_call_count`; you can just have a local variable (initially 1) that is used instead.

afd commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -64,13 +83,15 @@ namespace {
 struct GlobalData {
   InstanceMap instance_map;
   DeviceMap device_map;
+  std::atomic<uint64_t> current_draw_call = {};
```
Could do with a brief comment.

afd commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -80,6 +101,20 @@ struct GlobalData {
 };
 
+// Draw call state tracker is used to store the state of a draw call while
+// parsing the command buffers in vkQueueSubmit(...) -function.
```
Make it clear that there will be one of these per draw call being tracked.

afd commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -80,6 +101,20 @@ struct GlobalData {
+  std::vector<uint8_t> push_constant_data;
+
+  std::unordered_map<uint32_t, VkBuffer> bound_vertex_buffers;
+  std::unordered_map<uint32_t, VkDeviceSize> vertex_buffer_offsets;
```
It would probably be overkill to comment each of these, but the more interesting ones deserve a comment.

Is the order in which the fields appear a carefully thought through one?  Any reason why there's whitespace before the last two fields - i.e., are the first group of fields logically separated from the last two in some way?  If so that deserves a comment.

afd commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -122,21 +157,245 @@ void init_settings_if_needed() {
   }
 }
 
+//
```
I suggest removing the initial and trailing `//` lines.

afd replied:
OK, I see that you do this a lot, so it's a deliberate style thing.  @paulthomson do you have an opinion on whether we want to do this in the project?

paulthomson replied:
I have been adding fairly pointless comments like "Our vkEnumerateInstanceLayerProperties function." The reason is that these functions are standard Vulkan functions, so they don't really need a comment (we can still add a comment at the beginning of the implementation body). However, I feel all the Vulkan functions blend together a bit, so I added initial and trailing `//`. I suggest that we DON'T do this in general; just for Vulkan functions that don't really need a proper comment.

afd commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -122,21 +157,245 @@ void init_settings_if_needed() {
 }
 
+//
+// Adds the given vulkan command buffer command to the list of tracked commands.
```
"vulkan" -> "vulkan".

paulthomson replied:
Do you mean Vulkan?

afd commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -122,21 +157,245 @@ void init_settings_if_needed() {
+// from the draw call if the draw call is set to be captured via settings.
+//
+// TODO(ilkkasaa): This function is going to be huge. Consider making a separate
+// file for handling draw calls / creating amber files.
```
For the second (and following) line(s) of a TODO comment, please indent it by one space, so:

```
// TODO(ilkkkasaa): ....
//  file for ...
```

It shows up more nicely in CLion that way.

afd commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -122,21 +157,245 @@ void init_settings_if_needed() {
+                    uint32_t vertex_count, uint32_t first_instance,
+                    uint32_t instance_count) {
+  // Silence unused parameter warnings.
+  // TODO(ilkkasaa) remove these.
```
Elaborate: remove them when what?  When more functionality is added so that they are used?

afd commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -122,21 +157,245 @@ void init_settings_if_needed() {
+  DeviceData* device_data =
+      global_data->device_map.get(device_key(commandBuffer));
+
+  // Call the original function.
```
I suppose we should either use this comment uniformly (e.g. you haven't used it in vkCmdBindPipeline below), or omit it.

I think omitting would be fine as it's kind of obvious.

paulthomson replied:
I quite like it :-)

afd commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -122,21 +157,245 @@ void init_settings_if_needed() {
+
 //
 // Our vkQueueSubmit function.
+// Goes through all tracked commands in all of the submitted command buffers.
```
Say something about what it looks for when it goes through them.  In particular it's looking for draw calls.

paulthomson replied:
Maybe move this comment into the function body, as per my earlier comments about these standard Vulkan functions.

afd commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -122,21 +157,245 @@ void init_settings_if_needed() {
   DeviceData* device_data = global_data->device_map.get(device_key(queue));
 
-  // TODO(ilkkas) handle queue submits here.
+  for (uint32_t submit_idx = 0; submit_idx < submitCount; submit_idx++) {
```
I would have one comment per loop, to remind the reader of the structure of a Vulkan queue submission command.  E.g.: "Go through each queue submission.", then ...

afd commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -122,21 +157,245 @@ void init_settings_if_needed() {
 
-  // TODO(ilkkas) handle queue submits here.
+  for (uint32_t submit_idx = 0; submit_idx < submitCount; submit_idx++) {
+    for (uint32_t cmd_buffer_idx = 0;
```
some comment about the command buffers that make up a queue submission.

afd commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -122,21 +157,245 @@ void init_settings_if_needed() {
+          // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+          pSubmits[submit_idx].pCommandBuffers[cmd_buffer_idx];
+
+      // Ignore command buffers that are not tracked, i.e. command buffers that
+      // doesn't contain any interesting commands.
```
I like this comment.

paulthomson replied:
Me too. Except: that doesn't -> that don't

afd commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -122,21 +157,245 @@ void init_settings_if_needed() {
+
+      DrawCallStateTracker draw_call_state = {};
+      draw_call_state.queue = queue;
+      draw_call_state.command_buffer_handle = command_buffer_handle;
```
I would be inclined to lay the fields of DrawCallStateTracker out in the same order you populate them here.

afd commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -122,21 +157,245 @@ void init_settings_if_needed() {
+      }
+      DEBUG_ASSERT(device_data->command_buffers->count(command_buffer_handle));
+
+      DrawCallStateTracker draw_call_state = {};
```
It would be good to comment about what this is going to track.

afd reviewed (dismissed):
Looks good - my requests for changes are exclusively about comments and Google Style.

paulthomson commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -42,17 +50,28 @@ struct InstanceData {
+
+  // Tracked device data:
+
+  std::shared_ptr<
+      ProtectedMap<VkCommandBuffer, std::vector<std::shared_ptr<Cmd>>>>
+      command_buffers = std::make_shared<
+          ProtectedMap<VkCommandBuffer, std::vector<std::shared_ptr<Cmd>>>>();
```
Does this really need to be a `shared_ptr`? So far it looks likes the device owns this map. Could we just make it a ProtectedMap field?

ilkkasaa replied:
Oops, that is definitely a copy-paste error.

ilkkasaa replied:
@paulthomson Actually, looks like it has a reason to be a `shared_ptr`. If it's not a pointer, then the copy constructor of the `DeviceData` struct will be implicitly deleted, because the mutex used in the `ProtectedMap` has disabled copy constructor. And the copy constructor is used in `vkCreateDevice` function when updating the device map: `GetGlobalData()->device_map.put(device_key(*pDevice), device_data);` So the shared pointer is just for easier memory management. Of course we could use normal pointer and implement destructor to free the memory in the `DeviceData` struct or maybe even intercept `vkDestroyDevice` to properly free the `ProtectedMap` and delete the device from the `device_map` field.

paulthomson replied:
I have a commit that I can push to fix this. But I hesitate to push it in case you have unpushed changes.

paulthomson commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -122,21 +157,245 @@ void init_settings_if_needed() {
 
+//
+// Adds the given vulkan command buffer command to the list of tracked commands.
+// Creates a new list of commands for the given command buffer if it doesn't
+// already exist. Tracked commands will be parsed later when the command buffer
+// is submitted via vkQueueSubmit function.
+//
+void AddCommand(const DeviceData& device_data, VkCommandBuffer command_buffer,
```
Despite what I said about not needing to track stuff, I think we _should_ track vkAllocateCommandBuffers, vkResetCommandBuffer, vkResetCommandPool, and then also command pools! But this might be a fair amount of work, so perhaps just create an issue for it for now.

paulthomson replied:
Actually, we might be able to use a heuristic here; track when the command buffer has been submitted. If the app starts recording to such a command buffer then it must have been reset (otherwise I think the app must be violating the spec) so clear our vector before adding the new command.

ilkkasaa replied:
Right, that's a good point!

paulthomson commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -122,21 +157,245 @@ void init_settings_if_needed() {
+    device_data.command_buffers->put(command_buffer, std::move(empty_cmds));
+  }
+  device_data.command_buffers->get(command_buffer)
+      ->push_back(std::move(command));
```
Add a comment: The push_back call is not thread-safe, but this should be OK, as the Vulkan spec requires Vulkan applications to record commands without races.

paulthomson commented on src/VkLayer_GF_amber_scoop/include/vk_layer_amber_scoop/vulkan_commands.h:
```diff
@@ -0,0 +1,146 @@
+
+#include <cstdint>
+
+#include "vk_layer_amber_scoop/vk_deep_copy.h"
```
Please move the header files alongside `amber_scoop_layer.cc`. We can treat these an "internal" headers. My intention for include/ is for public headers, but I don't think we will have any of these for Vulkan layers.

paulthomson replied:
And you will just do `#include "vk_deep_copy.h"`

ilkkasaa replied:
@paulthomson Moving headers to same directory with .cc files causes some problems. As far as I know, "llvm-header-guard" expects header files to be located in a directory that has "include" somewhere in the path. Otherwise "llvm-header-guard" uses full path of the file to generate the header guard name, like this: `_HOME_ILKKAS_GIT_GF_LAYERS_SRC_VKLAYER_GF_AMBER_SCOOP_SRC_VK_DEEP_COPY_H`. Of course I can disable that check, but I think that's still a valuable check. Also "cpp-lint" wants the directory name when including header files: `#include "vulkan_commands.h"` causes error: `./src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:33:  Include the directory when naming .h files  [build/include_subdir]`. However that error can be avoided with `#include "./vulkan_commands.h"`

paulthomson replied:
Ah thanks for checking! Leave them as they are, but can you rename the vk_layer_amber_scoop/ to VkLayer_GF_amber_scoop/. I know this might violate some naming convention, but I think it will be the best solution.

paulthomson reviewed (commented):
I really like it! I left a few comments.

paulthomson commented on src/.clang-tidy:
```diff
@@ -14,6 +14,7 @@ Checks: >
   -fuchsia-default-arguments-calls,
   -fuchsia-statically-constructed-objects,
   -mpi-*,
+  -misc-non-private-member-variables-in-classes,
```
Please remove this before we merge.

paulthomson committed:
WIP: remove use of shared_ptr

paulthomson commented on src/gf_layers_layer_util/include/gf_layers_layer_util/util.h:
```diff
@@ -112,7 +113,11 @@ class ProtectedMap {
   // This is guaranteed for std::unordered_map.
   using InternalMapType = MapTemplate<KeyType, ValueType>;
-  using ThreadLocalCacheType = std::pair<KeyType, ValueType*>;
+
+  size_t count(KeyType key) const {
+    ScopedLock lock(mutex_);
+    return map_.count(key);
+  }
```
I just realised: we should probably not have a count function.

Use `get` if you want to maybe get something from the map.

Use `put` if you want to maybe put something in the map; check the bool return value to see if you actually did put it in the map.

ilkkasaa replied:
Yeah, I realised the same. The way I was using count() was to check if the map contained the given key, so I replaced `count(key)` with `contains(key)` that is just a shortcut and more readable form of `get(key) != nullptr`. 

And btw, functions in this file don't follow the function naming rules.

paulthomson replied:
Yeah my other concern is: `count` or `contains` encourages you to check something and then act upon the result, without holding the mutex throughout. It doesn't really matter in most cases, but I would prefer to not have `count` or `contains`. `get` and `put` kind of force you to attempt your action atomically.

Good point about the naming. I am not sure when to update these. Perhaps I should it now and you can rebase, as I have already made other changes, and you will probably need to rebase/merge anyway.

ilkkasaa replied:
I see, that makes sense.

ilkkasaa replied:
Oh, and go ahead with the naming updates.

paulthomson replied:
I think I will wait for naming updates, as I made some changes to util.h in this PR. Once we merge, I can quickly rename everything. For now, feel free to follow the existing style.

ilkkasaa committed:
Implement requested changes

Improves command buffer tracking.
Adds more comments about the code.
Changes Vulkan commands from classes to structs.

paulthomson commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -102,41 +193,283 @@ const std::array<VkLayerProperties, 1> kLayerProperties{{{
+      draw_call_state.queue = queue;
+
+      for (const auto& cmd : *command_buffer->GetCommandList()) {
+        if (auto* cmd_begin_renderpass = cmd->AsBeginRenderPass()) {
+          draw_call_state.current_render_pass =
+              &cmd_begin_renderpass->render_pass_begin;
+          draw_call_state.current_subpass = 0;
+        } else if (auto* cmd_draw = cmd->AsDraw()) {
+          HandleDrawCall(draw_call_state, *device_data, 0, 0,
+                         cmd_draw->first_vertex, cmd_draw->vertex_count,
+                         cmd_draw->first_instance, cmd_draw->instance_count);
+        } else if (auto* cmd_draw_indexed = cmd->AsDrawIndexed()) {
+          HandleDrawCall(
+              draw_call_state, *device_data, cmd_draw_indexed->first_index,
+              cmd_draw_indexed->index_count, 0, 0,
+              cmd_draw_indexed->first_index, cmd_draw_indexed->instance_count);
+        } else if (auto* cmd_bind_pipeline = cmd->AsBindPipeline()) {
+          // Currently we are interested in graphics pipelines only.
+          if (cmd_bind_pipeline->pipeline_bind_point ==
+              VK_PIPELINE_BIND_POINT_GRAPHICS) {
+            draw_call_state.graphics_pipeline = cmd_bind_pipeline->pipeline;
+          }
```
I quite like this because you can see all the cases, and it does feel like the code is quite "local" to this scope. However:

1. We could just use RTTI (`dynamic_cast`) instead of using a home-grown `AsX()` approach.
2. Use of RTTI (or any home-grown approach) is strongly discouraged by the Google C++ Style guide.
3. We could have a virtual method on Cmd called something like `ProcessSubmittedCommand` where we could put this code without needing to use RTTI. This last approach would also allow us to make all member variables private.

Let's discuss in tomorrow's meeting. I don't really mind either way, but we may as well discuss since we are so close to the meeting.

ilkkasaa committed:
Refactor draw call tracker and Cmd

Combines draw call state tracker and HandleDrawCall
function to a single class DrawCallTracker.

Removes the cast functions form Cmd-classes and adds new
ProcessSubmittedCommand-function.

Moves GlobalData, InstanceData, DeviceData and settings
struct definitions to a header file.

paulthomson commented on src/VkLayer_GF_amber_scoop/include/VkLayer_GF_amber_scoop/amber_scoop_layer.h:
```diff
@@ -0,0 +1,151 @@
+      contains_draw_calls_ = false;
+    }
+    // Set the flag if the command being added is a draw call.
+    if (cmd->GetKind() == Cmd::kDraw || cmd->GetKind() == Cmd::kDrawIndexed) {
```
Since this is the only use of `GetKind()`, can we remove the "kinds" completely and just have an `IsDrawCall` method? Or do you think we might end up needing "kinds" in the future?

ilkkasaa replied:
That's a good point. I guess we don't need the "kind" in the future. If we need it, it's not a big deal re-adding it.

paulthomson commented on src/VkLayer_GF_amber_scoop/include/VkLayer_GF_amber_scoop/draw_call_tracker.h:
```diff
@@ -0,0 +1,70 @@
+  struct DrawCallState {
+    VkCommandBuffer command_buffer_handle = nullptr;
+    VkQueue queue = nullptr;
+    VkRenderPassBeginInfo* current_render_pass = nullptr;
```
I don't think we should initialize these to nullptr; they can maybe sometimes not be pointers. We can just use {} instead. Since this is a struct, I have just removed all initializers, and we will intialize the struct with `{}` anyway.

paulthomson commented on src/VkLayer_GF_amber_scoop/include/VkLayer_GF_amber_scoop/amber_scoop_layer.h:
```diff
@@ -0,0 +1,151 @@
+// contains any draw calls. The class is not thread-safe, but it shouldn't
+// matter because the Vulkan spec requires Vulkan applications to record
+// commands without races.
+class CommandBufferTracker {
```
I think this class should go in its own file now, with `AddCommand` implemented in the .cc file. In the meantime, I think I will rename this to `CommandBufferData`.

paulthomson commented on src/VkLayer_GF_amber_scoop/include/VkLayer_GF_amber_scoop/amber_scoop_layer.h:
```diff
@@ -0,0 +1,151 @@
+
+  // Get reference to the list of captured commands. The list can't be modified
+  // directly. Use AddCommand() to add append the list.
+  [[nodiscard]] const std::vector<std::unique_ptr<Cmd>>* GetCommandList()
```
Apparently, we should avoid returning pointers in most cases; prefer a reference, unless the result is not always available.

paulthomson commented on src/VkLayer_GF_amber_scoop/include/VkLayer_GF_amber_scoop/amber_scoop_layer.h:
```diff
@@ -0,0 +1,151 @@
+// commands without races.
+class CommandBufferTracker {
+ public:
+  // Add command to the command list.
```
Function documentation comments must start with descriptive verbs in the indicative mood. E.g. "Adds |cmd|..." instead of "Add...". But then comments in function bodies are usually the other way around! I will change these.

paulthomson committed:
Some minor changes

paulthomson reviewed (commented):
This looks awesome!! I pushed a few minor changes and left a few comments. I have already addressed some of the comments in my push though!

paulthomson committed:
Fix check_iwyu.sh

ilkkasaa committed:
Move CommandBufferData to its own file

Moves CommandBufferData to its own .h and .cc files.
Removes "kind" from Cmd class and replaces it with
"IsDrawCall" function.
Some minor changes.

paulthomson reviewed (approved):
Nice!
